### ISSUE TITLE
Dead lock observed in BlockingWaitStrategy in Log 4J

### ISSUE THREAD

**Issue by SampathK** [LMAX-Exchange/disruptor#138]
We have seen this behavior in during high load. Where Logging Got Stropped and Application Went to not responsive state.
log4J Version : 2.2  Disruptor Version :  3.3.2
Ring Buffer Size : 128
Producer(Multiples Threads) and Consumer Threads(Single Thread As per Log 4J Configuration) Started Waiting on each other.

Here is the  one of the Trace from Thread Dump:

Producer : 
"[ACTIVE] ExecuteThread: '7' for queue: 'weblogic.kernel.Default (self-tuning)'" TIMED_WAITING
	sun.misc.Unsafe.park(Native Method)
	java.util.concurrent.locks.LockSupport.parkNanos(LockSupport.java:349)
	com.lmax.disruptor.MultiProducerSequencer.next(MultiProducerSequencer.java:136)
	com.lmax.disruptor.MultiProducerSequencer.next(MultiProducerSequencer.java:105)
	com.lmax.disruptor.RingBuffer.publishEvent(RingBuffer.java:444)
	com.lmax.disruptor.dsl.Disruptor.publishEvent(Disruptor.java:256)
	org.apache.logging.log4j.core.async.AsyncLogger.logMessage(AsyncLogger.java:285)
	org.apache.logging.log4j.spi.AbstractLogger.logMessage(AbstractLogger.java:722)
	org.apache.logging.log4j.spi.AbstractLogger.logIfEnabled(AbstractLogger.java:693)
        org.apache.logging.log4j.jcl.Log4jLog.debug(Log4jLog.java:81)

Consumer Thread : 
 
"AsyncLogger-1" waiting for lock      java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject@5d972983 WAITING
sun.misc.Unsafe.park(Native Method)
java.util.concurrent.locks.LockSupport.park(LockSupport.java:186)
java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject.await(AbstractQueuedSynchronizer.java:2043)
com.lmax.disruptor.BlockingWaitStrategy.waitFor(BlockingWaitStrategy.java:45)
com.lmax.disruptor.ProcessingSequenceBarrier.waitFor(ProcessingSequenceBarrier.java:55)
com.lmax.disruptor.BatchEventProcessor.run(BatchEventProcessor.java:123)
java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1145)
java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:615)
java.lang.Thread.run(Thread.java:744)

Is this is known issue which got already fixed in recent build ? 

Before raising this issue in Log4J thought of checking here since this dead lock related to LMAX implementation.

Can you please take a look into it ?, it seems to be critical for our application. Let me know if any other details required ?. Your Quick Support is much appreciated.

Code Snippet for BlockingWaitStrategy :
```
public final class BlockingWaitStrategy implements WaitStrategy
{
    private final Lock lock = new ReentrantLock();
    private final Condition processorNotifyCondition = lock.newCondition();

    @Override
    public long waitFor(long sequence, Sequence cursorSequence, Sequence dependentSequence, SequenceBarrier barrier)
        throws AlertException, InterruptedException
    {
        long availableSequence;
        if ((availableSequence = cursorSequence.get()) < sequence)
        {
           lock.lock();
            try
            {
                while ((availableSequence = cursorSequence.get()) < sequence)
                {
                    barrier.checkAlert();
                    processorNotifyCondition.await();
                }
            }
            finally
            {
                lock.unlock();
            }
        }

        while ((availableSequence = dependentSequence.get()) < sequence)
        {
            barrier.checkAlert();
        }

        return availableSequence;
    }

    @Override
    public void signalAllWhenBlocking()
    {
        lock.lock();
        try
        {
            processorNotifyCondition.signalAll();
        }
        finally
        {
            lock.unlock();
        }
    }
}
```

Regards,
Sakumar

**Comment by mikeb01** [comment]
What operating system and version are you running on?  What version of the JDK are you running on?

**Comment by SampathK** [comment]
Solaris Version 11 and JDK Version 7. Hope this helps you.

**Comment by SampathK** [comment]
Solaris Version 11 and JDK Version 7. Hope this helps you.

On Thu, Dec 10, 2015 at 12:15 AM, Michael Barker <notifications@github.com>
wrote:

> What operating system and version are you running on? What version of the
> JDK are you running on?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/LMAX-Exchange/disruptor/issues/138#issuecomment-163353845>
> .
>



-- 
Regards,
Sampath

**Comment by mikeb01** [comment]
Could you give the full versions of the OS and JDK (build numbers etc).

A couple of other things would also be useful.  1 a full stack trace.  2 the sequence numbers of the ring buffer and the batch event processor at the time that it failed.

**Comment by SampathK** [comment]
OS Version :
cat /etc/release
Oracle Solaris 11.2 X86

Java Version
java version "1.7.0_45"
Java(TM) SE Runtime Environment (build 1.7.0_45-b18)
Java HotSpot(TM) Server VM (build 24.45-b08, mixed mode)

1) Stack Trace :
Already provided thread dump trace, All Producer Threads Stuck at Same Point
2) Sequence Number : I don't think i can give you this number now, because Servers are Restarted to recover System. It is not a development System, Hence We cannot attach a debugger to it to reproduce.

I am new to LMAX Disruptor, if you could give me some idea about how this sequence number generated, i can write some client and try to reproduce it.

Regards,
Sakumar

**Comment by mikeb01** [comment]
The methods you need to get the sequence values from the Disruptor are:
Ring buffer sequence: `RingBuffer.getCursor()`
Consumer sequence: `BatchEventProcessor.getSequence().get()`

I'm not terribly familiar with Log4j 2.  They may have some monitoring code that can get that information.  You may also think about raising an enhancement request for Log4j to use the non-blocking calls to push to the Disruptor (`RingBuffer.tryPublishEvent()`), that way if the consumer is blocked the main application won't fail.

The bug report suggests that either the notification from the Condition variable was missed or the producer doesn't see the sequence value being updated from BatchEventProcessor.

**Comment by mikeb01** [comment]
The other action to do is if you run into this problem is to take a full heap dump.

**Comment by SampathK** [comment]
Yes We have Full Heap dump, it is more than 4 GB !!. I can't upload it.  I see its better you should suggest this enhancement to Log4J.  Because this may become critical for any System which uses Asynchronous Logging and i also don't see any work around to recover the system other than system restart :-( . This is definitely not acceptable by the customer which provides SLA for high availability systems.

**Comment by mikeb01** [comment]
Are you running 32 bit or 64 bit.

**Comment by mikeb01** [comment]
If you can find the RingBuffer and BatchEventProcessor objects in your heap dump.  There is a field of type Sequence called cursor on the AbstractSequencer that will contain the ring buffer sequence.  There is a field of the same type on the BatchEventProcessor called sequence that will contain the consumer sequence.

**Comment by SampathK** [comment]
Yes 64 bit

**Comment by SampathK** [comment]
Thanks !!. Let me check heap dump and see if i can provide the current sequence number.

**Comment by SampathK** [comment]
Got the Sequence Number from Heap BatchEventProcessor 764250528, RingBuffer  764250656

Attached Heap View for RingBuffer and BatchEventProcessor


[Heap_View_RingBuffer_BatchEventProcessor.docx](https://github.com/LMAX-Exchange/disruptor/files/58045/Heap_View_RingBuffer_BatchEventProcessor.docx)

**Comment by mikeb01** [comment]
Given the state of the sequence numbers this looks like the consumer missed the signal on the condition variable.  I'm not sure how this could happen.  The code has been working in our production system for around 5 years without seeing this issue, however we've never run is on Solaris.  I'm going to dig into Log4J next to see if they are doing anything strange. 

One other option is to add a periodic wakeup (e.g. every 100ms) to the BlockingWaitStrategy so that it won't get stuck inside await forever.

**Comment by SampathK** [comment]
Yes it is better to check from Log4J side. Is this issue can be common if
we have infinite producers and single consumer ?
Will this work around creates CPU busy waiting state ? .



On Fri, Dec 11, 2015 at 1:29 AM, Michael Barker <notifications@github.com>
wrote:

> Given the state of the sequence numbers this looks like the consumer
> missed the signal on the condition variable. I'm not sure how this could
> happen. The code has been working in our production system for around 5
> years without seeing this issue, however we've never run is on Solaris. I'm
> going to dig into Log4J next to see if they are doing anything strange.
>
> One other option is to add a periodic wakeup (e.g. every 100ms) to the
> BlockingWaitStrategy so that it won't get stuck inside await forever.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/LMAX-Exchange/disruptor/issues/138#issuecomment-163733664>
> .
>



-- 
Regards,
Sampath

**Comment by mikeb01** [comment]
It won't create a busy wait state, but will create a periodic wake up.

**Comment by SampathK** [comment]
OK is this issue will be fixed in any patch build

**Comment by SampathK** [comment]
Are you proposing below fix 
processorNotifyCondition.await(100L, TimeUnit.MILLISECONDS); instead of processorNotifyCondition.await(); in BlockWaitingStrategy

**Comment by mikeb01** [comment]
Yes, it's not really a fix, but a workaround.  I'm still confused about why the code would fail, the algorthm is pretty sound.  I'm wondering if it may be an issue on Solaris.

**Comment by SampathK** [comment]
Not sure why you are pointing Solaris,Can you tell me thread model used by Log4J Configuration like Multiple Producer and Single Consumer is it correct, did you verify RingBuffer Behavior in that use cases ?

**Comment by mikeb01** [comment]
The code Log4J looks reasonably sane.  They are using the DSL so that it assembles everything correct.  We use this wait strategy in a number of places and have been doing so for 5+ years.  The only thing different is Solaris.  Lock condition variables involve the kernel for handling park/unpark.  At some point I'll try an put together an isolated test that tries to trigger the problem.

**Comment by SampathK** [comment]
Ok Thanks. Since RCA is Unknown, i have decided to not to go with this work
around. I will wait for for your test results. If you require any other
details i will provide it.

On Fri, Dec 11, 2015 at 1:19 PM, Michael Barker <notifications@github.com>
wrote:

> The code Log4J looks reasonably sane. They are using the DSL so that it
> assembles everything correct. We use this wait strategy in a number of
> places and have been doing so for 5+ years. The only thing different is
> Solaris. Lock condition variables involve the kernel for handling
> park/unpark. At some point I'll try an put together an isolated test that
> tries to trigger the problem.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/LMAX-Exchange/disruptor/issues/138#issuecomment-163867238>
> .
>



-- 
Regards,
Sampath

**Comment by SampathK** [comment]
I have raised Jira issue to track from Log4j side

https://issues.apache.org/jira/browse/LOG4J2-1221

You can suggest any improvement from Log4J side based on your testing in above thread.

**Comment by SampathK** [comment]
Did you get a chance to simulate this in test environment. I checked with Log4J, they are also not recommending to use RingBuffer.tryPublishEvent() in log4j. We need a fix from LMAX Disruptor to avoid deadlock. Can you release a patch with timed waiting conditional lock ?

**Comment by mikeb01** [comment]
I've got back from my holiday break and will be digging into this further.  As an aside, if you switch to the Sleeping wait strategy (AsyncLoggerConfig.WaitStrategy=Sleep), then you will not have this problem.

This is the wait strategy that we employ for logging at LMAX as it has much less impact on the application threads (no lock wake up cost).  Although it does have a CPU cost to handle the regular thread wake ups.

**Comment by SampathK** [comment]
Welcome back Mike !!. I hope this will be a challenging issue to resolve it
in 2016.

Regarding Sleep wait I need to run test and confirm what is the deviation
in CPU usage ?

However I see its more important to find RCA for dead lock issue. If you
have any test prototype to reproduce, you can share it with me . I will run
the test,debug share the results in Solaris Systems.


On Thu, Jan 7, 2016 at 4:00 AM, Michael Barker <notifications@github.com>
wrote:

> I've got back from my holiday break and will be digging into this further.
> As an aside, if you switch to the Sleeping wait strategy
> (AsyncLoggerConfig.WaitStrategy=Sleep), then you will not have this problem.
>
> This is the wait strategy that we employ for logging at LMAX as it has
> much less impact on the application threads (no lock wake up cost).
> Although it does have a CPU cost to handle the regular thread wake ups.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/LMAX-Exchange/disruptor/issues/138#issuecomment-169482947>
> .
>



-- 
Regards,
Sampath

**Comment by mikeb01** [comment]
With regards to testing the deviation in performance you should run your normal performance tests and ensure that the change in performance profile for your application is not adversely affected.

The SleepingWaitStrategy is not too bad, but will consume 1-5% CPU when idle as it will do a frequent wake ups (~60µs on Linux, not sure about Solaris).  The only place where I've seen this cause issues is in heavily virtualised environments where CPU resource is overcommitted.  It is also probably not a good choice for a mobile application either.

**Comment by mikeb01** [comment]
So a workaround for this case is to use the TimeoutBlockingWaitStrategy and fix to make that the default has been push to Log4J.  I'll leave this issue open for the time being, until I get a chance to investigate the root cause.

**Comment by jamesacooper** [comment]
Deadlock with this wait strategy was reported a few years ago when using a very small ringbuffer size (4 in my case, which is small but the only the size constraints on the ringbuffer is that it's a power of 2).

**Comment by SampathK** [comment]
We are using 128 bytes

Regards,
Sampath Kumar

**Comment by mikeb01** [comment]
Workaround applied in Log4J, closing.

**Comment by SampathK** [comment]
Is RCA available for this issue ?  Log4j provided with work around timed
waiting

Regards,
Sampath Kumar

**Comment by mikeb01** [comment]
I am unable to replicate the problem on Linux and don't have access to a Solaris system for testing.

**Comment by SampathK** [comment]
Can you give me your testing prototype let me run it on Solaris and do some
analysis

Regards,
Sampath Kumar

**Comment by rajusu** [comment]
We have a similar issue occurring on in our environment. How can we get the latest version of log4j 2.6 or patches...? How can patches be applied.

Regards,
Sunil

**Comment by rajusu** [comment]
And we had this deadlock observed in Linux environment...

**Comment by mikeb01** [comment]
I've not been able to replicate it with the Disruptor in isolation, I'm going to try with log4j.  BTW, the simple workaround it is switch to the sleeping wait strategy instead of the default blocking one.

**Comment by GonzRon** [comment]
We are experiencing a similar issue with log4j (reported with LOG4J-1324).  The theory is that the original thread is exiting uncleanly (for unknown reasons, potentially an unhandled exception).  The 

The Daemon Thread Factory is attempting to spin up a new thread.  The new thread gets blocked on the lock which was never released by the original consumer thread.

On Left: Shows default consumer thread stack ("AsyncLoggerConfig-1")
On Right shows: New consumer thread ("AsyncLoggerConfig-2") attempting to be instantiated by daemon thread factory.

![2016-03-22_8-32-37](https://cloud.githubusercontent.com/assets/18006895/13951861/b210bf6a-f009-11e5-9781-5203df801b69.jpg)

**Comment by mikeb01** [comment]
I've not been able to replicate the log4j deadlock issue locally, with both a test on top of the Disruptor or Log4J.  However I've added an experimental fix to the Disruptor.  If it detects that the ring buffer is full it will re-signal the wait strategy, this should unstick the consumer if it happened to miss a signal.

It is available as version 3.3.5-rc2, you can get the source from the releases page or via the git repo.  You'll need to compile it and drop the resulting binary (from the build/libs) directory in place.

It would be really useful if one of the users experiencing issues could try this version out.  If this doesn't fix the issue, then it is unlikely to be a concurrency issue and more like a Exception knocking out the consumer thread.

**Comment by mikeb01** [comment]
I fairly convinced that this is not an issue with the Disruptor, but something to do with the set up within Log4J.  As @GonzRon suggests, they appear to tear down and recreate the Disruptor instance and it is possible that the EventHandler thread is stuck on an old condition object.

To help diagnose this, there is a toString method on the Disruptor that will dump information around the BlockingWaitStrategy.  It will include the object id of the WaitStrategy's condition object and the object id of any condition that the event handler thread is currently waiting on.  These should match.  If they don't then there is some sort of issue.  This will only work if you use the non-deprecated Disruptor constructor that uses a ThreadFactory.

**Comment by mikeb01** [comment]
I'm going to close due to no further activity. In commit 5c5a33850e8b7259f1ccf8bd7a213bf16035e309 I added support for re-signalling the listeners if the producer encounters a full ring buffer, so if the signal happened to be lost somehow (not convinced that it can be) then it will be on the next attempted publication.

This code was included in release 3.3.5.

**Comment by jainhj** [comment]
Hi,

I have got the same issue with YieldingWaitStrategy on linux machine with disruptor-3.3.6. Can someone please suggest best workaround or solution.

regards,
Honi Jain

**Comment by mikeb01** [comment]
Start by upgrading to 3.4.2.

**Comment by mikeb01** [comment]
Also make sure you are on the latest version of log4j2.

**Comment by jainhj** [comment]
sure thanks. I will try that and let you know.

**Comment by jainhj** [comment]
Hi,

I moved to latest version of disruptor but I am on log4j1. However, the problem is not coming while logging. Do I Still need to upgrade to latest version of log4j2. There is no logging in the code path in which the above problem came.

Can someone please advise.

**Comment by mikeb01** [comment]
You placed your comment in a thread based on an issue with Log4j. If the problem you are encountering is not related to Log4j, please can you raise a separate issue.

**Comment by jainhj** [comment]
sure thanks
